src/actions/breakpoints/breakpointPositions.js

```diff
@@ -5,55 +5,65 @@
 // @flow
 
 import { isOriginalId, originalToGeneratedId } from "devtools-source-map";
+import { uniqBy } from "lodash";
 
-import { getSourceFromId, hasBreakpointPositions } from "../../selectors";
+import {
+  getSourceFromId,
+  hasBreakpointPositions,
+  getBreakpointPositionsForSource
+} from "../../selectors";
 
 import type { ThunkArgs } from "../../actions/types";
 import { getOriginalLocation } from "../../utils/source-maps";
+import { makeBreakpointId } from "../../utils/breakpoint";
 
 const requests = new Map();
 
-async function mapLocations(generatedLocations, state, source, sourceMaps) {
+async function mapLocations(generatedLocations, { sourceMaps }) {
   return Promise.all(
     generatedLocations.map(async generatedLocation => {
-      const location = await getOriginalLocation(
-        generatedLocation,
-        source,
-        sourceMaps
-      );
+      const location = await getOriginalLocation(generatedLocation, sourceMaps);
 
       return { location, generatedLocation };
     })
   );
 }
 
-function convertToList(results, sourceId) {
+function filterByUniqLocation(positions) {
+  return uniqBy(positions, ({ location }) => makeBreakpointId(location));
+}
+
+function convertToList(results, source) {
+  const { id, url } = source;
   const positions = [];
 
   for (const line in results) {
     for (const column of results[line]) {
-      positions.push({ line: Number(line), column: column, sourceId });
+      positions.push({
+        line: Number(line),
+        column: column,
+        sourceId: id,
+        sourceUrl: url
+      });
     }
   }
 
   return positions;
 }
 
-async function getBreakpointPositions(
-  sourceId,
-  { client, dispatch, getState, sourceMaps }
-) {
-  let source = getSourceFromId(getState(), sourceId);
+async function _setBreakpointPositions(sourceId, thunkArgs) {
+  const { client, dispatch, getState, sourceMaps } = thunkArgs;
+  let generatedSource = getSourceFromId(getState(), sourceId);
```

Comment:
I think naming this `generatedSource` when it could be original seems potentially confusing. Maybe we can leave the existing variable and declare a new one for the eventual generated source?